### DISCUSSION TITLE
Implement getStaticProps using a different field other than id?

### DISCUSSION THREAD

**Discussion by richleach** [vercel/next.js#25620] [Help] [answered]
Setting up a blog on my next.js site and I went through the exercise on 
https://nextjs.org/learn/basics/dynamic-routes/implement-getstaticprops
I had to reference Nader's video at https://www.youtube.com/watch?v=bQ1Giqn5G38&t=1192s since my datastore is being hosted on Amazon Web Service's Amplify. 
The problem is currently my urls are looking like http://localhost:3000/blog/22ed553b-5590-4b09-a155-cc3bed664c30 when I would like them to look more like http://localhost:3000/blog/this-is-my-url (I need SEO friendly urls).
I'm able to get my pages working as expected but unfortunately I'm passing the generic 'id' value in my urls, and I'd really like to use a 'url' field that I specifically set up to use for SEO purposes.
I tried changing [id].js to [url].js (as 'url' is actually the name of the column in my DynamoDB), then in the [url].js I changed the references from 'id' to 'url', but I kept getting a JSON parse error....  
Is there a section in the docs that describes how I would use a different value other than id?
Thank you in advance,
Rich 

//my /pages/blog/index.js file that lists all of the blog posts to the user with the id in the url (I want to replace the id with my seo-pretty url value)

import {useState, useEffect} from 'react'
import Link from 'next/link'
import { DataStore } from '@aws-amplify/datastore';
import { Post } from '../../models';
import Moment from 'react-moment';

export default function index() {
    const [posts, setPosts] = useState([])
    useEffect(() => {
        fetchPosts()
        async function fetchPosts(){
            const postData = await DataStore.query(Post)
            setPosts(postData)
        }
    }, [])
    return (
        <div>
            <h2>My Posts</h2>
            {
                posts.map(post => (
                    <Link href={`/blog/${post.id}`} key={post.id}>
                        <a>
                            <p>{post.title} <br />Posted on <Moment date={post.postDate} format="LL" /> by {post.author} <br />
                            Categories: {post.categories}
                            </p>
                        </a>
                    </Link>
                ))
            }
        </div>
    )
}

//my [id].js file using getStaticPaths/getStaticProps. This works using the id, but I can't seem to get it working when I use a custom url value instead of id
import { withSSRContext} from 'aws-amplify'
import { Post } from '../../models'
import { useRouter } from 'next/router'
import Markdown from 'react-markdown'
import Moment from 'react-moment';

export default function PostComponent({post}) {
    const router = useRouter()
    console.log(post)
    if(router.isFallback) {
        return <div>Loading....</div>
    }
    return(
        <div>
            <h1>{post.title}</h1>
            <p>{post.body}
            Posted on <Moment date={post.postDate} format="LL" /> by {post.author} <br />
            Categories: {post.categories}</p>
        </div>
    )
}

//this retrieves a list of posts, which will get rendered into a list of paths
export async function getStaticPaths(req) {
    const { DataStore } = withSSRContext(req)
    const posts = await DataStore.query(Post)
    const paths = posts.map(post => ({params: {id: post.id}}))
//console.log(paths)
    return {
        paths,
        fallback: true
    }
}

//this will retrieve the data for the selected post
export async function getStaticProps(req) {
    const { DataStore } = withSSRContext(req)
    const { params } = req
    const { id } = params

    
    const post = await DataStore.query(Post, id)

    return {
        props: {
            post: JSON.parse(JSON.stringify(post))
        },
        revalidate: 100
    }
}

**Comment by wwwdepot** [comment, marked as answer]
Check your getStaticPaths/getStaticProps `await DataStore.query(Post, id)` usage. Passing a string in 2nd params only query against id column. If you want to query other columns, you need to use a criteria function (see 2nd link).

DataStore types: https://aws-amplify.github.io/amplify-js/api/classes/datastore.html#query
Amplify docs: https://docs.amplify.aws/lib/datastore/data-access/q/platform/js#predicates

**Reply by richleach** [reply]
You were spot on! I'm posting my final solution below, thank you for your help.

**Reply by richleach** [reply]
I marked your reply as the correct answer for crediting purposes, and below I uploaded my modified code in the current, corrected state.

**Comment by richleach** [comment]
1. In getStaticProps I had to add a filter to return only the item where the passed in 'url' matched the item's 'url'
2. Because the returned value is actually an array of 1 object, I had to modify my returned output to reference all values ordinally (even though there's only 1 returned value). Before I was using {post.title} now I'm using {post[0].title}.
3. Don't forget to change the name of your file, before the query was using id so the filename was [id].js ; now i'm using the url column in my table so the filename is now [url].js


import { withSSRContext} from 'aws-amplify'
import { Post } from '../../models'
import { useRouter } from 'next/router'
import Markdown from 'react-markdown'
import Moment from 'react-moment';

export default function PostComponent({post}) {
    const router = useRouter()
    console.log(post)
    if(router.isFallback) {
        return <div>Loading....</div>
    }
    return(
        <divq>
            <h1q>{post[0].title}</h1q>
            <pq>{post[0].body}
            Posted on <Moment date={post[0].postDate} format="LL" /> by {post[0].author} <brq />
            Categories: {post[0].categories}</pq>
        </divq>
    )
}

//this retrieves a list of posts, which will get rendered into a list of paths
export async function getStaticPaths(req) {
    const { DataStore } = withSSRContext(req)
    const posts = await DataStore.query(Post)
    const paths = posts.map(post => ({params: {url: post.url}}))
//console.log(paths)
    return {
        paths,
        fallback: true
    }
}

//this will retrieve the data for the selected post
export async function getStaticProps(req) {
    const { DataStore } = withSSRContext(req)
    const { params } = req
    const { url } = params

    //const post = await DataStore.query(Post, id)  
    const post = await DataStore.query(Post, c => c.url("eq", url))

    return {
        props: {
            post: JSON.parse(JSON.stringify(post))
        },
        revalidate: 100
    }
}
